### PR TITLE
Collapse date joins

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderDate.java [hunk 1]
```diff
@@ -56,7 +60,18 @@ public Predicate addPredicate(String theResourceName,
...
+		boolean newJoin = false;
+		if (myJoinMap == null) {
+			myJoinMap = new HashMap<>();
+		}
+		String key = theResourceName + " " + theParamName;
+
+		Join<ResourceTable, ResourceIndexedSearchParamDate> join = myJoinMap.get(key);
+		if (join == null) {
+			join = createJoin(SearchBuilderJoinEnum.DATE, theParamName);
+			myJoinMap.put(key, join);
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderDate.java]: if it wasn't for the newJoin flag, I'd suggest a putIfAbsent here...

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderDate.java [hunk 2]
```diff
@@ -77,7 +92,14 @@ public Predicate addPredicate(String theResourceName,
 		}
 
 		Predicate orPredicates = myBuilder.or(toArray(codePredicates));
-		myQueryRoot.addPredicate(orPredicates);
+
+		if (newJoin) {
+			Predicate identityAndValuePredicate = combineParamIndexPredicateWithParamNamePredicate(theResourceName, theParamName, join, orPredicates);
+			myQueryRoot.addPredicate(identityAndValuePredicate);
+		} else {
+			myQueryRoot.addPredicate(orPredicates);
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderDate.java]: The fact this change was so easy is a testament to how well designed your solution was in the first place.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java [hunk 3]
```diff
@@ -3258,6 +3258,73 @@ public void testSearchWithDate() {
...
+		IIdType obsId1;
+		{
+			Observation obs1 = new Observation();
+			obs1.setIssuedElement(new InstantType("2020-06-06T12:00:00Z"));
+			obs1.setEffective(new InstantType("2019-06-06T12:00:00Z"));
+			obsId1 = myObservationDao.create(obs1).getId().toUnqualifiedVersionless();
+		}
+		IIdType obsId2;
+		{
+			Observation obs1 = new Observation();
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java]: calling this obs1 is confusing.  I'd just call them both obs

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java [hunk 4]
```diff
@@ -3258,6 +3258,73 @@ public void testSearchWithDate() {
...
+			myCaptureQueriesListener.clear();
+			SearchParameterMap params = new SearchParameterMap();
+			params.setLoadSynchronous(true);
+			params.add("issued", new DateParam("ge2020-06-05"));
+			params.add("issued", new DateParam("lt2020-06-07"));
+			List<IIdType> patients = toUnqualifiedVersionlessIds(myObservationDao.search(params));
+			assertThat(patients, contains(obsId1));
+			String searchQuery = myCaptureQueriesListener.getSelectQueriesForCurrentThread().get(0).getSql(true, true);
+			ourLog.info("Search query:\n{}", searchQuery);
+			assertEquals(searchQuery, 1, StringUtils.countMatches(searchQuery.toLowerCase(), "join"));
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java]: This is beautiful.  That captures query listener is the best!

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java [hunk 5]
```diff
@@ -3258,6 +3258,73 @@ public void testSearchWithDate() {
...
+			List<IIdType> patients = toUnqualifiedVersionlessIds(myObservationDao.search(params));
+			assertThat(patients, contains(obsId1));
+			String searchQuery = myCaptureQueriesListener.getSelectQueriesForCurrentThread().get(0).getSql(true, true);
+			ourLog.info("Search query:\n{}", searchQuery);
+			assertEquals(searchQuery, 2, StringUtils.countMatches(searchQuery.toLowerCase(), "join"));
+			assertEquals(searchQuery, 2, StringUtils.countMatches(searchQuery.toLowerCase(), "hash_identity"));
+			assertEquals(searchQuery, 4, StringUtils.countMatches(searchQuery.toLowerCase(), "sp_value_low"));
+		}
+	}
+
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java]: Consider adding a couple of period searches for good measure.  I'd recommend a test with two periods that overlap.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderDate.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/predicate/PredicateBuilderDate.java]: Heh yup

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java]: Good call, done

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java (hunk 5 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchNoFtTest.java]: done

**Pull request merged**